### PR TITLE
[sharktank] Add chunk attention support for Llama 4 model

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
This commit adds support for chunk attention for the prefill step of the Llama 4 model.

The code is adopted from https://github.com/nod-ai/shark-ai/tree/users/sogartar/llama4.

Authored-by: Boian Petkantchin <Boian.Petkantchin@amd.com>
Co-authore-by: Vivek Khandelwal <vivekkhandelwal1424@gmail.com>

### CODE REVIEW

File: sharktank/sharktank/models/llm/llm.py [hunk 1]
```diff
@@ -144,7 +144,7 @@ def prefill(
         # [bs, batch_seq_len]
         tokens: Union[torch.Tensor, ReplicatedTensor],
         *,
-        # [[1, 1, batch_seq_len, batch_seq_len] x self.config.pipeline_parallelism_size]
+        # [[bs|1, 1, batch_seq_len, batch_seq_len] x self.config.pipeline_parallelism_size]
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/llm.py:147]: Why does attention_mask itself change to `bs | 1` for it's first dim?

File: sharktank/sharktank/layers/causal_llm.py [hunk 2]
```diff
@@ -142,6 +142,56 @@ def attention_mask(
...
+        start_index = 0
+        end_index = batch_seq_len
+        chunked_boolean_attention_mask = self.create_boolean_chunked_attention_mask(
+            attention_chunk_size=self.config.attention_chunk_size,
+            # TODO: handle decode step
+            start_index=start_index,
+            end_index=end_index,
+        )
+
+        return torch.where(
```
Comment (Alex-Vasile) [inline sharktank/sharktank/layers/causal_llm.py:158]: This won't work for sharded models. is the plan to come back and extend to sharded afterwards?

File: sharktank/sharktank/layers/causal_llm.py [hunk 3]
```diff
@@ -142,6 +142,56 @@ def attention_mask(
         ).to(dtype)
         return numeric_mask.to(self.device)
 
+    def chunked_attention_mask(self, attention_mask: torch.Tensor) -> torch.Tensor:
```
Comment (Alex-Vasile) [inline sharktank/sharktank/layers/causal_llm.py:145]: Type hint is incorrect. `attention_mask` is `torch.Tensor | ReplicatedTensor`.

File: sharktank/sharktank/models/llm/llm.py [hunk 4]
```diff
@@ -162,16 +162,31 @@ def prefill(
...
+            ), "TODO: implement pipeline-parallel chunked attention"
+            chunked_attention_mask = [self.chunked_attention_mask(attention_mask[0])]
+
         # Iterate over attention blocks.
         for block_idx, block in enumerate(self.attn_blocks):
             if block_idx == 0:
                 self.trace_tensor(f"llama.attn_block.{block_idx}.input", h)
+            use_chunked_attention = (
+                self.config.attention_chunk_size is not None
+                and (block_idx + 1) % 4 != 0
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/models/llm/llm.py:178]: Right. In the config `LlamaModelConfig` we must add filed that drives what layers are chunked and what are not.

[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/models/llm/llm.py:178]: Then when converting from a Hugging Face config we would populate this field.

[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/models/llm/llm.py:178]: I think with this change we would be good to land this.

File: sharktank/sharktank/layers/causal_llm.py (hunk 2 above)
Comment (sogartar) [reply sharktank/sharktank/layers/causal_llm.py:158]: I say we worry about sharding later.

File: sharktank/sharktank/models/llm/llm.py [hunk 5]
```diff
@@ -162,16 +163,34 @@ def prefill(
...
 
+        if self.config.attention_chunk_size is not None:
+            chunked_attention_mask = [
+                ops.replicate(
+                    self.chunked_attention_mask(attention_mask[0]),
+                    count=len(self.cache.pipeline_to_device_map),
+                    devices=self.cache.pipeline_to_device_map[pipeline],
+                )
+                for pipeline in range(self.cache.pipeline_count)
+            ]
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llm/llm.py:174]:
```suggestion
        if self.config.attention_chunk_size is not None:
            chunked_attention_mask = [
                ops.replicate(
                    self.chunked_attention_mask(attention_mask[pipeline]),
                    count=len(self.cache.pipeline_to_device_map[pipeline]),
                    devices=self.cache.pipeline_to_device_map[pipeline],
                )
                for pipeline in range(self.cache.pipeline_count)
            ]
```
This will require further work for sharded/pipeline models, but will be fine for now for unsharded.

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 6]
```diff
@@ -212,6 +212,10 @@ class LlamaModelConfig:
     # the program and not.
     static_tables: bool = True
 
+    attention_chunk_size: Optional[int] = None
+
+    chunked_attention_layers: Optional[set[int]] = None
+
```
Comment (Alex-Vasile) [inline sharktank/sharktank/layers/configs/llm_configs.py:218]: Need descriptions

File: sharktank/sharktank/models/llm/llm.py [hunk 7]
```diff
@@ -162,16 +163,34 @@ def prefill(
...
+
         # Iterate over attention blocks.
         for block_idx, block in enumerate(self.attn_blocks):
             if block_idx == 0:
                 self.trace_tensor(f"llama.attn_block.{block_idx}.input", h)
+            use_chunked_attention = (
+                self.config.attention_chunk_size is not None
+                and (block_idx + 1) % len(self.config.chunked_attention_layers) != 0
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/models/llm/llm.py:182]: This does not seem correct. It should be that `self.config.chunked_attention_layers` is a set of all layer/block indexes that use chunked attention. It should be
```
use_chunked_attention = self.config.chunked_attention_layers is not None and block_idx in self.config.chunked_attention_layers
```

[the remark this answers is not in the document]
Comment (Alex-Vasile) [reply sharktank/sharktank/models/llm/llm.py:182]: That version makes sense

**Review by sogartar — APPROVED** [review]
